File: src/main/scala/org/ximo/finkinaction/scala/dataset/DataSetMapPartitionApp.scala
package org.ximo.finkinaction.scala.dataset

import java.lang

import org.apache.flink.api.common.functions.MapPartitionFunction
import org.apache.flink.api.scala.{ExecutionEnvironment, _}
import org.apache.flink.util.Collector

/**
  *
  *
  * @author xikl
  * @date 2019/8/5
  */
object DataSetMapPartitionApp {

  def main(args: Array[String]): Unit = {
    val env = ExecutionEnvironment.getExecutionEnvironment
    //    mapPartition1(env)
        mapPartition2(env)
//    mapPartition3(env)
  }


  def mapPartition1(env: ExecutionEnvironment): Unit = {
    env.fromCollection(1 to 100)
      .setParallelism(3)
      .mapPartition((values: Iterator[Int], collector: Collector[Int]) => {
        var c = 0
        for (s <- values) {
          c += 1
        }
        collector.collect(c)
      }).print()
  }

  /**
    * 34
    * 33
    * 33
    * 分组 并行去弄
    *
    * @param env
    */
  def mapPartition2(env: ExecutionEnvironment): Unit = {
    env.fromCollection(1 to 100)
      .setParallelism(3)
      .mapPartition(values => List(values.length))
      .print()
  }

  def mapPartition3(env: ExecutionEnvironment): Unit = {
    val partitionMapper = new MapPartitionFunction[Int, Int] {
      override def mapPartition(values: lang.Iterable[Int], out: Collector[Int]): Unit = {
        var c = 0
        values.forEach(_ => c += 1)
        out.collect(c)
      }
    }

    val partitionMapper2: MapPartitionFunction[Int, Int] = (values, out) => {
      var c = 0
      values.forEach(_ => c += 1)
      out.collect(c)
    }

    env.fromCollection(1 to 100)
      .setParallelism(3)
      .mapPartition(partitionMapper2)
      .print()
  }


}

File: src/main/java/org/ximo/finkinaction/java/dataset/DataSetMapPartitionApp.java
package org.ximo.finkinaction.java.dataset;

import org.apache.flink.api.common.functions.MapPartitionFunction;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.util.Collector;

import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

/**
 * @author xikl
 * @date 2019/8/4
 */
public class DataSetMapPartitionApp {

    public static void main(String[] args) throws Exception {
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        List<String> data = Stream.iterate(1, i -> i + 1)
                .map(Object::toString)
                .limit(100)
                .collect(toList());

        //34
        //33
        //33
        env.fromCollection(data)
                .setParallelism(3)
                .mapPartition(new PartitionCounter())
                .print();

        // 有点意思 首先lambda中要写明类型 否则无法正常编写
        env.fromCollection(data)
                .setParallelism(3)
                .mapPartition((Iterable<String> values, Collector<Long> out) -> {
                    long c = 0;
                    for (String s : values) {
                        c++;
                    }
                    out.collect(c);
                })
                .returns(Types.LONG)
                .print();

    }

    public static class PartitionCounter implements MapPartitionFunction<String, Long> {

        @Override
        public void mapPartition(Iterable<String> values, Collector<Long> out) {
            long c = 0;
            for (String s : values) {
                c++;
            }
            out.collect(c);
        }
    }
}
